[AIRFLOW-3370] Es handler

rhwang10 opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow-3370](https://issues.apache.org/jira/browse/AIRFLOW-3370) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-3370

### Description

Add additional options and documentation for using the ElasticsearchTaskHandler.

The easiest and most foolproof way to implement a logging solution is to write to the standard output streams (stdout and stderr). In the Kubernetes ecosystem, given that pods are killed and restarted constantly, this implies persistent storage is a requirement for log history preservation.

For the webserver and scheduler components of Airflow, logging to standard output streams is built in. However, when tasks are executed in Celery, workers will fork off child processes to execute tasks concurrently. Before the child process ends, it makes a call to the Airflow task logger, and a task log file is written to the file system.

This potentially causes several problems with the Airflow on top of Kubernetes architecture. Given that Airflow has a constant stream of log output, running an Airflow environment using Celery in a Kubernetes cluster requires large amounts of memory resources. As such, when memory resources are exceeded, either 1) worker pods are often evicted by Kubernetes, or 2) worker output stalls and tasks pile up without completing. A current best-case workaround is to run a sidecar container that tails the `stdout` and `stderr` streams to read logs from the filesystem of the worker node and then output those logs to its own standard output. However, this becomes a scaling issue when running multiple deployments, and is unsustainable for massive Airflow deployments.

The options that are added in this PR look to circumvent the need for persistent volumes in worker nodes when running Airflow on Kubernetes. Workers will no longer need to be Stateful Sets and can instead be Deployments. 

The `elasticsearch_write_stdout` flag in `airflow.cfg` will allow the child process to write its log output to the parent process standard output stream.
The `elasticsearch_json_format` flag in `airflow.cfg` allows additional optional JSON configuration for task instances based on the `logging` module LogRecord attributes. This must be used in conjunction with the `elasticsearch_record_labels` configuration. 

A potential use case for these options is precisely when setting up a log monitoring stack, such as EFK (Elasticsearch FluentD Kibana), without requiring persistent volumes. A FluentD daemon listening on every node is awaiting log output on the standard output stream. With the `write_stdout` flag set, it can capture the task log information that has been executed on child processes from the parent process standard output. Using the `json_format` field, FluentD can be configured to filter and specify log records, without needing to parse the standard Airflow log formatted record, before sending it off to a destination, such as Elasticsearch, where it can stored and handled independent of Airflow on Kubernetes.

If either of these options are NULL, or not set, the `es_task_handler.py` will function exactly as before, and will only have read functionality. The options in this PR simply provide more functionality and options to users.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

#### handler close
- test_close_stdout_logs
- test_close_closed_stdout_logs
- test_close_no_mark_end_stdout_logs
- test_close_with_no_handler_stdout_logs
- test_close_with_no_stream_stdout_logs

#### handler read
- test_read_stdout_logs
- test_read_nonexistent_log_stdout_logs
- test_read_raises_stdout_logs_json_true
- test_read_timeout_stdout_logs
- test_read_with_empty_metadata_stdout_logs
- test_read_with_none_metadata_stdout_logs

#### handler render_log_id and set_context
- test_render_log_id_json_true
- test_set_context_stdout_logs

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

schnie commented on airflow/utils/log/es_task_handler.py:
```diff
@@ -68,6 +100,76 @@ def __init__(self, base_log_folder, filename_template,
+        if self.handler is not None:
+            self.handler.flush()
+
+    def _process_taskInstance(self, ti):
```
@rhwang10 should this be `_process_task_instance` to stay consistently snake cased?

schnie commented on airflow/utils/log/es_task_handler.py:
```diff
@@ -19,14 +19,41 @@
+    def __init__(self, processedTask=None):
+        super(JsonFormatter, self).__init__()
+        self.processedTask = processedTask
+    def format(self, record):
```
@rhwang10 maybe add a newline here?

schnie commented on airflow/utils/log/es_task_handler.py:
```diff
@@ -19,14 +19,41 @@
 
+RECORD_LABELS = ['asctime', 'levelname', 'filename', 'lineno', 'message']
+
+class parentStdout():
```
@rhwang10 should this class name be capitalized?

schnie commented:
@rhwang10 how do you configure human vs json format?

rhwang10 commented:
@schnie setting the `json_format` env should be enough- it picks between the json formatter and the airflow default formatter- will add that back in

schnie commented:
@rhwang10 should it be added to the default cfg file like the other env?

rhwang10 commented:
@schnie yep, should be under `[elasticsearch]`

schnie commented:
I don't see that one includes in this PR. I just see write_stdout.

rhwang10 commented:
fixed linting errors + moved record_labels to config file

schnie commented:
@rhwang10 I think this is looking pretty good. I haven't tried it yet, but how do you feel about it now? Is there anything you think is left to ship this to apache and into our stable branches?

Does the existing elasticsearch handler have any tests? Think there's any functions we could write a test for? The PR will show that we decreased code coverage when we send it.

Maybe @andscoop could give it a quick look as well.

rhwang10 commented:
@schnie i'm reasonably happy with it too, def want to review it with you/andy in more depth. it should definitely have some test written for it- i'll look into it

schnie commented:
@rhwang10 looking great! Anything left ya think?

rhwang10 commented:
@schnie think it's almost ready, some last things: 1) last checkbox says to add some documentation if new functionality is added, and 2) some commit messages don't have [AIRFLOW-3370] as a prefix

rhwang10 commented:
Added docs for Elasticsearch (surprised there was nothing about the es_handler at all yet), rewrote all the commits, listed the tests- looks like we can submit tomorrow by EOD

schnie commented on dags/test_dag.py:
```diff
@@ -1,43 +0,0 @@
-# -*- coding: utf-8 -*-
```
@rhwang10 isn't this file part of the official Apache project?

rhwang10 replied:
Oops, you're right

schnie commented on airflow/config_templates/__init__.py:
```diff
@@ -16,3 +16,4 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+from . import airflow_local_settings
```
@rhwang10 whats this addition doing?

rhwang10 replied:
I think this was an addition to get the dev setup up and running- had to do with how the PYTHONPATH was exported in the dockerfile- we can remove it

rhwang10 commented:
@schnie i think i've covered most of the bases- am I missing anything?

schnie commented:
@rhwang10 I think it looks good. You may want to squash this down to just a handful of meaningful commits if possible.

rhwang10 commented:
@schnie all tests pass 👍 think it's good to go
